File: pipeline/compose.py
"""Paint-density thickness mask + simple bump composer.

paint_density_mask: detects where paint is piled up by measuring local
variance of luminance, with a luminance weighting so bright impasto
regions register strongest.

compose_bump: multiplies the stroke field by the thickness mask, recenters,
and rescales the 99th percentile to `output_amplitude`.
"""

from __future__ import annotations

import numpy as np
import cv2


def paint_density_mask(
    luminance_norm: np.ndarray,
    gamma: float = 1.5,
    floor: float = 0.0,
) -> np.ndarray:
    """Per-pixel paint-density multiplier in [0, 1].

    Detects WHERE paint is piled up by measuring LOCAL VARIANCE of luminance
    in a moderate window. Key properties:
      * A single sharp edge contributes contrast^2 * (edge_len/window) to
        the variance -> diluted by window size -> low density.
      * A cluster of fine brushwork contributes ~constant variance
        independent of window size -> high density.
    This keeps silhouette/contour edges from registering as "thick paint",
    while preserving genuine textured regions (hair, brush patches).

    Then luminance-weighted: bright textured regions (real impasto catches
    light) get boosted (×1.5 at white), dark textured regions (often jpeg
    noise in shadows) get suppressed (×0.5 at black).

    Optional `floor` lifts the mask to give the canvas a baseline texture
    even where raw density is zero.

    Parameters
    ----------
    luminance_norm : np.ndarray (H, W)
        Luminance normalized to [0, 1].
    gamma : float
        Power applied to the final density. >1 emphasizes peaks.
    floor : float in [0, 1)
        Minimum mask value: ``mask = floor + (1 - floor) * raw``.
    """
    lum = np.clip(luminance_norm.astype(np.float32), 0.0, 1.0)
    var_radius = 25
    ks = 2 * var_radius + 1
    mean = cv2.boxFilter(lum, ddepth=-1, ksize=(ks, ks), normalize=True)
    sq_mean = cv2.boxFilter(lum * lum, ddepth=-1, ksize=(ks, ks), normalize=True)
    variance = np.clip(sq_mean - mean * mean, 0.0, None)
    density = np.sqrt(variance)  # std-dev: linear in contrast amplitude
    # Mild spatial smoothing so the mask is locally stable.
    density = cv2.GaussianBlur(density, (17, 17), 5.0)
    # Robust normalization to [0, 1] using a high percentile.
    p95 = float(np.percentile(density, 95))
    if p95 > 1e-8:
        density = density / p95
    density = np.clip(density, 0.0, 1.0)
    # Luminance weighting: impasto catches light; shadow noise is suppressed.
    lum_factor = 0.5 + lum  # range [0.5, 1.5]
    density = np.clip(density * lum_factor, 0.0, 1.0)
    density = np.power(density, max(gamma, 1e-3)).astype(np.float32)
    if floor > 0.0:
        density = (floor + (1.0 - floor) * density).astype(np.float32)
    return density


def compose_bump(
    stroke_field: np.ndarray,
    thickness: np.ndarray,
    output_amplitude: float = 0.5,
    recenter: bool = True,
) -> np.ndarray:
    """Combine the stroke field with the thickness mask.

    bump = stroke_field * thickness

    If ``recenter`` is True (default), then subtract the mean and rescale so
    the 99th percentile of |bump| equals ``output_amplitude``, with final
    clip to ``[-output_amplitude, +output_amplitude]``.

    Parameters
    ----------
    stroke_field : np.ndarray (H, W)
        Anisotropic stroke field, roughly in [-1, 1].
    thickness : np.ndarray (H, W)
        Per-pixel multiplier in [0, 1].
    output_amplitude : float
        Target |bump| 99th percentile after recentering.
    recenter : bool
        If True, subtract mean and rescale; else return raw product.
    """
    if stroke_field.shape != thickness.shape:
        raise ValueError(
            "shape mismatch: stroke "
            f"{stroke_field.shape} vs thickness {thickness.shape}"
        )

    bump = (
        stroke_field.astype(np.float32) * thickness.astype(np.float32)
    ).astype(np.float32)

    if not recenter:
        return bump

    bump = bump - float(np.mean(bump))
    p99 = float(np.percentile(np.abs(bump), 99))
    target = max(float(output_amplitude), 1e-6)
    if p99 > 1e-8:
        bump = bump * (target / p99)
    bump = np.clip(bump, -target, target)
    return bump.astype(np.float32)

File: web/src/compose.js
// Paint-density 厚度 mask + 简单 bump 合成器。
//
// 移植自 pipeline/compose.py。

import { boxFilter, gaussianBlur, percentileAbs, mean } from './filters.js';

/**
 * paint_density_mask: 基于局部方差检测 "哪里有颜料堆积"。
 *
 * @param {Float32Array} luminanceNorm  ∈ [0, 1]
 * @param {number} W
 * @param {number} H
 * @param {object} [opts]
 * @param {number} [opts.gamma=1.5]
 * @param {number} [opts.floor=0.0]
 * @returns {Float32Array}
 */
export function paintDensityMask(luminanceNorm, W, H, opts = {}) {
    const gamma = opts.gamma ?? 1.5;
    const floor = opts.floor ?? 0.0;
    const N = W * H;

    // clip 到 [0, 1]
    const lum = new Float32Array(N);
    for (let i = 0; i < N; i++) {
        const v = luminanceNorm[i];
        lum[i] = v < 0 ? 0 : (v > 1 ? 1 : v);
    }

    // 局部方差：mean & E[x^2]，box filter (radius=25, win=51)
    const varRadius = 25;
    const lumSq = new Float32Array(N);
    for (let i = 0; i < N; i++) lumSq[i] = lum[i] * lum[i];

    const meanArr = boxFilter(lum, W, H, varRadius);
    const sqMean = boxFilter(lumSq, W, H, varRadius);
    const density = new Float32Array(N);
    for (let i = 0; i < N; i++) {
        let varVal = sqMean[i] - meanArr[i] * meanArr[i];
        if (varVal < 0) varVal = 0;
        density[i] = Math.sqrt(varVal); // std-dev：与对比度幅度线性
    }

    // 轻微空间平滑，让 mask 局部稳定。Python: GaussianBlur(17, 17, 5.0)
    const densitySmooth = gaussianBlur(density, W, H, 5.0);

    // 用第 95 百分位归一化到 [0, 1]
    const p95 = percentileAbs(densitySmooth, 95);
    const out = new Float32Array(N);
    if (p95 > 1e-8) {
        for (let i = 0; i < N; i++) {
            let v = densitySmooth[i] / p95;
            if (v < 0) v = 0;
            else if (v > 1) v = 1;
            out[i] = v;
        }
    }
    // 否则全 0

    // luminance 加权 (亮的 ×1.5, 暗的 ×0.5)，再 clip 到 [0, 1]
    for (let i = 0; i < N; i++) {
        let v = out[i] * (0.5 + lum[i]);
        if (v < 0) v = 0;
        else if (v > 1) v = 1;
        out[i] = v;
    }

    // gamma
    const g = Math.max(gamma, 1e-3);
    if (g !== 1.0) {
        for (let i = 0; i < N; i++) out[i] = Math.pow(out[i], g);
    }

    // floor
    if (floor > 0.0) {
        const f = floor;
        const span = 1.0 - f;
        for (let i = 0; i < N; i++) out[i] = f + span * out[i];
    }

    return out;
}

/**
 * compose_bump: bump = stroke * thickness, 居中并按第 99 百分位重缩放到 outputAmplitude。
 *
 * @param {Float32Array} strokeField
 * @param {Float32Array} thickness
 * @param {number} W
 * @param {number} H
 * @param {object} [opts]
 * @param {number} [opts.outputAmplitude=0.5]
 * @param {boolean} [opts.recenter=true]
 * @returns {Float32Array}
 */
export function composeBump(strokeField, thickness, W, H, opts = {}) {
    const outputAmplitude = opts.outputAmplitude ?? 0.5;
    const recenter = opts.recenter ?? true;
    const N = W * H;
    if (strokeField.length !== N || thickness.length !== N) {
        throw new Error('strokeField / thickness shape mismatch');
    }

    const bump = new Float32Array(N);
    for (let i = 0; i < N; i++) bump[i] = strokeField[i] * thickness[i];

    if (!recenter) return bump;

    const m = mean(bump);
    for (let i = 0; i < N; i++) bump[i] -= m;

    const p99 = percentileAbs(bump, 99);
    const target = Math.max(outputAmplitude, 1e-6);
    if (p99 > 1e-8) {
        const s = target / p99;
        for (let i = 0; i < N; i++) bump[i] *= s;
    }
    // clip
    for (let i = 0; i < N; i++) {
        const v = bump[i];
        bump[i] = v < -target ? -target : (v > target ? target : v);
    }
    return bump;
}
